server/src/service/agenda.ts

```diff
@@ -0,0 +1,172 @@
+import type { User } from "@prisma/client";
+import * as schema from "biseo-interface/agenda";
+import { prisma } from "@/db/prisma";
+import { BiseoError } from "@/lib/error";
+import { BiseoServer } from "@/types/socket";
+
+export const retrieveAll = async (
+  {}: schema.RetrieveAll,
+  user: User
+): Promise<schema.RetrieveAllCb> => {
+  const agendaDbRes = await prisma.agenda.findMany({
+    where: { deletedAt: null },
+    include: {
+      voters: true,
+      choices: {
+        select: {
+          id: true,
+          name: true,
+          users: true,
+        },
+      },
+    },
+  });
+  const res = agendaDbRes.map((agenda): schema.Agenda => {
+    const userVotable = agenda.voters.some((v) => v.userId == user.id);
+    //TODO: possible optimization
+    const userVoted = userVotable
+      ? agenda.choices.find((c) => c.users.some((u) => u.userId == user.id))
+          ?.id || null
+      : null;
+    const commonField = {
+      id: agenda.id,
+      title: agenda.title,
+      content: agenda.content,
+      resolution: agenda.subtitle,
+      voters: {
+        voted: agenda.choices.reduce(
+          (acc, choice) => acc + choice.users.length,
+          0
+        ),
+        total: agenda.voters.length,
+      },
+    };
+    return !agenda.startAt
+      ? {
+          ...commonField,
+          status: "preparing",
+          user: {
+            votable: userVotable,
+          },
+        }
+      : !agenda.endAt
+      ? {
+          ...commonField,
+          status: "ongoing",
+          user: {
+            votable: userVotable,
+            voted: userVoted,
+          },
+          choices: agenda.choices.map((choice) => {
+            return {
+              id: choice.id,
+              name: choice.name,
+            };
+          }),
+        }
+      : {
+          ...commonField,
+          status: "terminated",
+          user: {
+            votable: userVotable,
+            voted: userVoted,
+          },
+          choices: agenda.choices.map((choice) => {
+            return {
+              id: choice.id,
+              name: choice.name,
+              count: choice.users.length,
+            };
+          }),
+        };
+  });
+  if (!res) throw new BiseoError("failed to retrieve agenda");
+
+  return res;
+};
+
+export const vote = async (
+  { choiceId, agendaId }: schema.Vote,
+  io: BiseoServer,
+  user: User
+) => {
+  // validation
+  const isUserVotable =
+    (await prisma.userAgendaVotable.count({
+      where: {
+        userId: user.id,
+        agendaId: agendaId,
+      },
+    })) > 0;
+  const isChoiceInAgenda =
+    (await prisma.choice.count({
+      where: {
+        id: choiceId,
+        agendaId: agendaId,
+      },
+    })) > 0;
+  if (isUserVotable && isChoiceInAgenda) {
```

Comment:
이부분을 nesting하는 것보다는 early return을 사용하여 nesting을 줄이고 가독성을 향상시킬 수 있을 것 같아요. 
```suggestion
  if (!await prisma.userAgendaVotable.count({
    where: {
      userId: user.id,
      agendaId: agendaId,
    },
  })) throw new BiseoError("Do not have permissions to vote to this agenda");
  
  if (!await prisma.choice.count({
    where: {
      id: choiceId,
      agendaId: agendaId,
    },
  })) throw new BiseoError("Already voted for this agenda");
  
```

이와 같이 개선하면 Client 쪽에도 조금 더 상세한 에러를 줄 수 있을 것 같네요!

그때 톡방에 올린 적이 있었던 [이 영상](https://www.youtube.com/watch?v=CFRhGnuXG-4)의 리팩토링 방법을 실제 적용해 볼 수 있는 사례인 것 같아요!

Reply:
저 count하는 쿼리를 if문의 조건으로 그냥 적으면 왜 검사한건지 알기 어려울 것 같아요.
지금처럼 변수명을 명확히 적어주거나, 주석을 남겨서 어떤 조건을 검사하고 싶었는지 보여주면 좋을 것 같아요.

Reply:
앗 저 리뷰의 의도는 그걸 말하려는게 아니라
```
if (조건 & 조건) {
  // 메인 로직
}
```
형태보다 
```
if (!조건) throw;
if (!조건) throw;

// 메인 로직
```
과 같은 식으로 코드를 구성했을 때 로직을 더 잘 정리할 수 있다는 뜻이었어요.

변수명을 유지하는 방식으로 구조를 그렇게 변경하면 이렇게 되겠네요.
```suggestion
  const isUserVotable = !!(await prisma.userAgendaVotable.count({
    where: {
      userId: user.id,
      agendaId: agendaId,
    },
  }));
  if (!isUserVotable) throw new BiseoError("Do not have permissions to vote to this agenda");

  const isChoiceInAgenda = !!(await prisma.choice.count({
    where: {
      id: choiceId,
      agendaId: agendaId,
      NOT: { users: { some: { userId: user.id } } },
    },
  }));
  if (!isChoiceInAgenda) throw new BiseoError("Invalid choice");


```

Reply:
앗 그리고 전 리뷰에서는 발견하지 못했는데 유저가 이미 투표했는지에 대한 validation이 추가되어야 할 것 같네요
```typescript
  const userVoted = !!(await prisma.userChoice.count({
    where: {
      userId: user.id,
      choiceId: choiceId,
    }
  }));
  if (userVoted) throw new BiseoError("Already voted for this agenda");
```

Reply:
아 그렇네요. 이미 투표했는지도 확인해야겠네요.